File: Python/Queue/649. Dota2 Senate.py
""" 649. Dota2 Senate 
Difficulty: Medium
Submission Link: https://leetcode.com/problems/dota2-senate/submissions/1387464896/ """

class Solution(object):
    def predictPartyVictory(self, senate):
        R = deque()
        D= deque()
        l = len(senate)-1
        for s in range(len(senate)):
            if senate[s] == "R":
                R.append(s)
            else:
                D.append(s)

        while R and D:
            r= R.popleft()
            d = D.popleft()
            if r < d :
                R.append(r+l)
            else:
                D.append(d+l)
        
        return "Radiant" if R else "Dire"

        
        

File: C++/Queue/649. Dota2 Senate.cpp
/* 649. Dota2 Senate 
Difficulty: Medium
Submission Link: https://leetcode.com/problems/dota2-senate/submissions/1387465106/ */

class Solution {
public:
    string predictPartyVictory(string s) {
        queue<int>r ; 
        queue<int>d ; 
        int n = s.length() ; 
        for(int i = 0 ; i < n ; i++){
            (s[i] == 'R') ? r.push(i) : d.push(i) ;
        }
        while(!r.empty() && !d.empty()){
            int x , y ; 
            x = r.front(); 
            y = d.front(); 
            r.pop() ; 
            d.pop() ; 
            (x < y) ? r.push(n + x) : d.push(n + y) ; 
        }
        return !r.empty() ? "Radiant" : "Dire" ; 
    }
};